design-documents/graph-ql/nullability.md

```diff
@@ -0,0 +1,216 @@
+
+```graphql
+type Query {
+    # If a client queries for both fields, and one of them fails,
+    # the client will receive no data
+    product(id: ID): Product!
+    category(id: ID): Category!
+}
+```
+
+### Primary keys (id field) should always be non-nullable
+
+It very rarely makes sense to have a resource that _can_ have an ID but might not.
+
+IDs are extremely important for caching in most GraphQL clients, so it's worthwhile to be safe here.
+
+```graphql
+type Product {
+    id: ID! # Rarely makes sense for this to be nullable
+}
+```
+
+### Scalars should be non-nullable when required in backing data source
+
+Scalars form the leaves of a response in GraphQL, and are the bits of data the client really cares about. Because of this, we should strive to make scalars non-nullable when possible.
+
+The exceptions to this rule are:
+
+- A scalar field _should_ be nullable if it can be empty in the application/backing data source
+- A scalar field _should_ be nullable if it's likely to come from a different service than its parent object (fairly rare)
+
+
+### Consider the Parent Type
+
+If you're not dealing with an `id` field or a top-level `Query` field, the most important question to ask is:
+
+**Is the parent object still usable if this field has an error?**
+
+#### Example: Parent still usable with field error
+
+```graphql
+type Product {
+    # Recommended products are not critical data on a product page, and a UI can represent
+    # a product safely without related products, so we keep the field nullable
+    recommended_products: ProductRecommendations
+}
+```
+
+#### Example: Parent not usable with field error
+
+```graphql
+type Product {
+    # A user would not be able to add a product to the cart from the Product
+    # details page if this field fails, because it may have required options.
+    # We make the field's type non-nullable
+    options: ProductOptions!
+}
+```
+
+### List Types
+
+List fields have [2 distinct forms of nullability](http://spec.graphql.org/draft/#sec-Combining-List-and-Non-Null):
+
+1. Field nullability (same as all other fields)
+2. List item nullability (whether a list can have `null` items inside it)
+
+These forms can be composed together in various ways:
+```graphql
+type Example {
+    foo1: [Foo]   # foo1 can be null or a list. If list, can have nulls in it
+    foo2: [Foo!]! # foo2 must be a list, and every entry must be a 'Foo'
+    foo3: [Foo!]  # foo3 can be a list or null. If list, every item must be a 'Foo'
+    foo4: [Foo]!  # foo4 must be a list, and the list can have null values
+}
+```
+
+To decide whether the _field_ should be nullable, use the other recommendations provided in this document.
+
+When deciding whether _List items_ should be nullable, the most important question to ask is:
+
+**"Is the parent object still usable if at least one item in this list has an error?"**
```

Comment:
This is very inconvenient, requiring each iterator to check for null which is quite a PITA for functional programming often used in React with TypeScript. If a child has a GraphQL error (and I hope the goal is to have 0 cases of that), I would much prefer the entire field to get an error than to have to work around nullability within every single array where it never happens that an item is actually null.